### PR TITLE
Use a persistent connection per thread instead of connecting anew for each event

### PR DESCRIPTION
**Pull request by samstokes**
Prior to this PR, libhoney-rb is making a brand new HTTPS connection to api.honeycomb.io before sending each event. This is extremely inefficient, incurring both a TCP connection setup and a TLS handshake for each event. Each outbound connection also consumes kernel resources (and continues to do so for around a minute while the connection is in TCP `CLOSE_WAIT`). Finally this may be causing a problem observed by a potential customer who found that libhoney-rb was exhausting their port address translations.

The "http" gem we are using [supports persistent connections](https://github.com/httprb/http/wiki/Persistent-Connections-%28keep-alive%29) using the HTTP/1.1 header `Connection: Keep-Alive`, so let's use them. Each thread in the sending thread pool (default 10 threads) gets its own persistent connection (*) and sends all events it processes on that same connection.

It's hard to obtain a representative benchmark, but my initial results suggest that for a moderately-loaded, multi-threaded Rails app that is not CPU-limited (i.e. CPU is not pegged at 100% during the test), this reduces the CPU overhead of libhoney-rb by a factor of 2-3, as well as reducing the number of open outbound connections from hundreds to 10.

The http gem takes care of managing the persistent connection. We just receive, and reuse, a persistent `Client` object, but under the hood it sets up a `Connection`, detects if it has timed out or been interrupted and reconnects if needed. I've verified this behaviour through local testing and inspection.

(*) Actually, its own set of persistent connections. Because we support specifying the API host per-event, we actually keep a persistent connection _per API host_. This is a nuance that probably matters to zero people, but it also moves us toward the code structure we'll need to support batching (#1).

### CODE REVIEW

File: lib/libhoney/transmission.rb [hunk 1]
```diff
@@ -72,16 +83,22 @@ def send_loop
           # nothing consuming the queue).
           response = Response.new(:error => error)
         ensure
-          response.duration = Time.now - before
-          response.metadata = e.metadata
+          if response
```
Comment (toshok) [inline lib/libhoney/transmission.rb]: this `if` makes me think it's possible to end up here without a valid response.  That doesn't seem possible?

File: lib/libhoney/transmission.rb [hunk 2]
```diff
@@ -72,16 +83,22 @@ def send_loop
...
+            response.duration = Time.now - before
+            response.metadata = e.metadata
+          end
         end
 
         begin
-          @responses.enq(response, !@block_on_responses)
+          @responses.enq(response, !@block_on_responses) if response
```
Comment (toshok) [inline lib/libhoney/transmission.rb]: same here.  how do we end up here without a `response`?

File: lib/libhoney/transmission.rb (hunk 1 above)
Comment (samstokes) [reply lib/libhoney/transmission.rb]: The short version is that I observed this happening during testing (with extra debug output added, in a moderate load test, hitting edge cases like the queue filling up, etc). Unfortunately I didn't track down exactly why.

The long version is no, it shouldn't be possible to end up here without a `response`, but because it's an `ensure` block (and given I saw it in the wild) it's worth being extra paranoid. (It's outside the `rescue` so exceptions here would kill the thread, and also an exception in the `ensure` block could mask an original exception.)

I'm a little puzzled what could cause it - an exception not matched by the `rescue` pattern would do it, but `Exception` should match all exceptions; `Response.new` throwing an exception would do it, but that's a PORO whose constructor doesn't throw things.

I think I want to refactor this code anyway - especially if I do end up implementing batching - and I'd definitely want to clean up the error handling along the way. The current patchwork of `rescue` blocks does not well express the intention of "this is our thread pool and we do not want these threads to terminate until we ask them to".

File: lib/libhoney/transmission.rb (hunk 2 above)
Comment (samstokes) [reply lib/libhoney/transmission.rb]: If we made it to the `ensure` block above without a response then it'll still be nil here :(
